internal/pipeline/plugin.go

```diff
@@ -19,40 +22,92 @@ var (
 //
 // Standard caveats apply - see the package comment.
 type Plugin struct {
-	Name   string
+	Source string
 	Config any
 }
 
 // MarshalJSON returns the plugin in "one-key object" form, or "single string"
-// form (no config, only plugin name).
+// form (no config, only plugin name). Plugin sources are marshalled into "full"
+// form.
 func (p *Plugin) MarshalJSON() ([]byte, error) {
 	// NB: MarshalYAML (as seen below) never returns an error.
 	o, _ := p.MarshalYAML()
 	return json.Marshal(o)
 }
 
 // MarshalYAML returns the plugin in either "one-item map" form, or "scalar"
-// form (no config, only plugin name).
+// form (no config, only plugin name). Plugin sources are marshalled into "full"
+// form.
 func (p *Plugin) MarshalYAML() (any, error) {
 	if p.Config == nil {
-		return p.Name, nil
+		return p.FullSource(), nil
 	}
 
 	return map[string]any{
-		p.Name: p.Config,
+		p.FullSource(): p.Config,
 	}, nil
 }
 
+// FullSource attempts to canonicalise Source. If it fails, it returns Source
+// unaltered. Otherwise, it resolves sources in a manner described at
+// https://buildkite.com/docs/plugins/using#plugin-sources.
+func (p *Plugin) FullSource() string {
+	if p.Source == "" {
+		return ""
+	}
+
+	// Looks like an absolute or relative file path.
+	if strings.HasPrefix(p.Source, "/") || strings.HasPrefix(p.Source, ".") || strings.HasPrefix(p.Source, `\`) {
+		return p.Source
+	}
+
+	u, err := url.Parse(p.Source)
+	if err != nil {
+		return p.Source
```

Comment:
I wonder if there's a way that we can carry across that `FullSource` had to fall back to being identical on a particular input. Then, when it comes time to sign the plugin, we can feed a warning to the job logs. It's perhaps not necessary right now, but I would want such a thing before we announce this is released.

Reply:
I think it's a bit tricky. Some inputs have to fall back because neither the pipeline-uploading agent nor the backend know what local or vendored plugins will be present. And some inputs must fall back because there needs to be a way to override the assumption that the plugin is in GitHub (and usually in buildkite-plugins). Failing to parse as a URL / URI could be a mistake, but I have the suspicion that some edge case will fall into it.